### PR TITLE
MES-4255 - Cat C View Test Result

### PR DESCRIPTION
**Pull request by TomBillingtonUK**
## Description
- Created a Cat C View Test Result
- Refactored Rekey Details Card to remove some duplicated code. I'll keep sneaking these small refactors in for the ones not yet done

## Checklist

- [x] PR title includes the JIRA ticket number
- [x] Branch is rebased against the latest develop
- [x] Code has been tested manually
- [ ] PR link added to JIRA ticket
- [ ] One review from each scrum team
- [ ] Squashed commit contains the JIRA ticket number

### CODE REVIEW

File: src/pages/view-test-result/cat-c/__tests__/view-test-result.cat-c.page.spec.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        expect(result.testCentreName).toBe('Example Test Centre');
+      });
+      it('should return null when there is no test result', () => {
+        const result: ExaminerDetailsModel = component.getExaminerDetails();
+        expect(result).toBeNull();
+      });
+    });
+    describe('getHeaderDetails', () => {
+      it('should return the correct data', () => {
+        component.testResult = categoryBETestResultMock;
```
Comment (rossfauldsbjss) [inline src/pages/view-test-result/cat-c/__tests__/view-test-result.cat-c.page.spec.ts:148]: TODO comment or CATC result mock.

File: src/pages/view-test-result/cat-c/components/business-details-card/__tests__/business-details-card.spec.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      });
+  }));
+
+  describe('Class', () => {
+    describe('shouldHideCard', () => {
+      it('should return true if all data is present', () => {
+        expect(component.shouldHideCard()).toEqual(true);
+      });
+      it('should return false if all data has been provided', () => {
+        const data: CatBEUniqueTypes.Candidate = {
```
Comment (rossfauldsbjss) [inline src/pages/view-test-result/cat-c/components/business-details-card/__tests__/business-details-card.spec.ts:47]: TODO comment or CATCUniqueTypes

**Review by rossfauldsbjss — APPROVED** [review]
I made some comments about TODOS or use CAT C type.
I'm assuming another PR will replace the CatBE types with CAT C.
Happy to approve on that basis.

File: src/pages/view-test-result/cat-c/__tests__/view-test-result.cat-c.page.spec.ts [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+import { MockComponent } from 'ng-mocks';
+import { TestDetailsCardComponent } from '../../components/test-details-card/test-details-card';
+import { RekeyDetailsCardComponent } from '../../components/rekey-details-card/rekey-details';
+import { RekeyReasonCardComponent } from '../../components/rekey-reason-card/rekey-reason';
+import { ExaminerDetailsCardComponent } from '../../components/examiner-details-card/examiner-details';
+import { By } from '@angular/platform-browser';
+import { ExaminerDetailsModel } from '../../components/examiner-details-card/examiner-details-card.model';
+import { TestDetailsModel } from '../../components/test-details-card/test-details-card.model';
+import { VehicleDetailsCardComponent } from '../components/vehicle-details-card/vehicle-details-card';
+import { categoryBETestResultMock } from '../../../../shared/mocks/cat-be-test-result.mock';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/__tests__/view-test-result.cat-c.page.spec.ts:30]: Can we have a generic category mock or do we need to create a Cat C one?

File: src/pages/view-test-result/cat-c/components/business-details-card/__tests__/business-details-card.spec.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
+
+import { ComponentFixture, async, TestBed } from '@angular/core/testing';
+import { IonicModule, Config, Platform } from 'ionic-angular';
+import { ConfigMock, PlatformMock } from 'ionic-mocks';
+import { MockComponent } from 'ng-mocks';
+import { DataRowComponent } from '../../../../../../components/common/data-row/data-row';
+import { DataRowCustomComponent } from '../../../../../../components/common/data-row-custom/data-row-custom';
+import { BusinessDetailsCardComponent } from '../business-details-card';
+import { CatBEUniqueTypes } from '@dvsa/mes-test-schema/categories/BE';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/components/business-details-card/__tests__/business-details-card.spec.ts:10]: Add a TODO comment?

File: src/pages/view-test-result/cat-c/components/debrief-card/__tests__/debrief-card.spec.ts [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  DrivingFaultsBadgeComponent,
+} from '../../../../../../components/common/driving-faults-badge/driving-faults-badge';
+import { DataRowComponent } from '../../../../../../components/common/data-row/data-row';
+import { DataRowCustomComponent } from '../../../../../../components/common/data-row-custom/data-row-custom';
+import { DataRowWithListComponent } from '../../../../components/data-row-with-list/data-list-with-row';
+import { FaultsDataRowComponent } from '../../../../components/faults-data-row/faults-data-row';
+import { VehicleChecksDataRowComponent } from '../../../../components/vehicle-checks-data-row/vehicle-checks-data-row';
+import { FaultSummaryProvider } from '../../../../../../providers/fault-summary/fault-summary';
+import { FaultCountProvider } from '../../../../../../providers/fault-count/fault-count';
+import { CatBEUniqueTypes } from '@dvsa/mes-test-schema/categories/BE';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/components/debrief-card/__tests__/debrief-card.spec.ts:24]: Add TODO comment?

File: src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import { QuestionResult } from '@dvsa/mes-test-schema/categories/Common';
+
+@Component({
+  selector: 'debrief-card',
+  templateUrl: 'debrief-card.html',
+})
+export class DebriefCardComponent {
+
+  @Input()
+  data: CatBEUniqueTypes.TestData;
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts:27]: Add TODO comment?

File: src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  public getSeriousFaults(): FaultSummary[] {
+    return this.faultSummaryProvider.getSeriousFaultsList(this.data, TestCategory.BE);
+  }
+
+  public getDangerousFaults(): FaultSummary[] {
+    return this.faultSummaryProvider.getDangerousFaultsList(this.data, TestCategory.BE);
+  }
+
+  public getDrivingFaultCount(): number {
+    return this.faultCountProvider.getDrivingFaultSumCount(TestCategory.BE, this.data);
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts:97]: Add TODO comments for TestCategory.BE?

File: src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+import { get } from 'lodash';
+import { flattenArray } from '../../../view-test-result-helpers';
+import { TestCategory } from '@dvsa/mes-test-schema/categories/common/test-category';
+import { CatBEUniqueTypes } from '@dvsa/mes-test-schema/categories/BE';
+import {
+  DataRowListItem,
+  ViewTestResultLabels,
+  TestRequirementsLabels,
+} from '../../../components/data-row-with-list/data-list-with-row.model';
+import { manoeuvreTypeLabels } from '../../../../../shared/constants/competencies/catbe-manoeuvres';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts:13]: Add TODO comment?

File: src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
+import { Component, Input } from '@angular/core';
+import { get } from 'lodash';
+import { flattenArray } from '../../../view-test-result-helpers';
+import { TestCategory } from '@dvsa/mes-test-schema/categories/common/test-category';
+import { CatBEUniqueTypes } from '@dvsa/mes-test-schema/categories/BE';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts:6]: Add TODO comment?

File: src/pages/view-test-result/cat-c/components/vehicle-details-card/__tests__/vehicle-details-card.spec.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
+
+import { ComponentFixture, async, TestBed } from '@angular/core/testing';
+import { IonicModule, Config } from 'ionic-angular';
+import { ConfigMock } from 'ionic-mocks';
+import { VehicleDetailsCardComponent } from '../vehicle-details-card';
+import { MockComponent } from 'ng-mocks';
+import { DataRowComponent } from '../../../../../../components/common/data-row/data-row';
+import { DataRowCustomComponent } from '../../../../../../components/common/data-row-custom/data-row-custom';
+import { VehicleDetails } from '@dvsa/mes-test-schema/categories/Common';
+import { CatBEUniqueTypes } from '@dvsa/mes-test-schema/categories/BE';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/components/vehicle-details-card/__tests__/vehicle-details-card.spec.ts:11]: Add TODO comment?

File: src/pages/view-test-result/cat-c/components/vehicle-details-card/vehicle-details-card.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import { Component, Input } from '@angular/core';
+import { CatBEUniqueTypes } from '@dvsa/mes-test-schema/categories/BE';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/components/vehicle-details-card/vehicle-details-card.ts:3]: Add TODO comment?

File: src/pages/view-test-result/cat-c/view-test-result.cat-c.page.module.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+import { IonicPageModule } from 'ionic-angular';
+import { ViewTestResultCatCPage } from './view-test-result.cat-c.page';
+import { SearchProvider } from '../../../providers/search/search';
+import { CompressionProvider } from '../../../providers/compression/compression';
+import { EffectsModule } from '@ngrx/effects';
+import { ViewTestResultAnalyticsEffects } from '../view-test-result.analytics.effects';
+import { ComponentsModule } from '../../../components/common/common-components.module';
+import { FaultSummaryProvider } from '../../../providers/fault-summary/fault-summary';
+import { ViewTestResultComponentsModule } from '../components/view-test-result.components.module';
+import { ViewTestResultCatBEComponentsModule } from './components/view-test-result.cat-be.components.module';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/view-test-result.cat-c.page.module.ts]: Add TODO comment?

File: src/pages/view-test-result/cat-c/view-test-result.cat-c.page.ts [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+import { SaveLog } from '../../../modules/logs/logs.actions';
+import { LogHelper } from '../../../providers/logs/logsHelper';
+import { QuestionProvider } from '../../../providers/question/question';
+import { TestCategory } from '@dvsa/mes-test-schema/categories/common/test-category';
+import { HttpResponse } from '@angular/common/http';
+import { TestDetailsModel } from '../components/test-details-card/test-details-card.model';
+import { ExaminerDetailsModel } from '../components/examiner-details-card/examiner-details-card.model';
+import { ViewTestHeaderModel } from '../components/view-test-header/view-test-header.model';
+import { CatBEUniqueTypes } from '@dvsa/mes-test-schema/categories/BE';
+import { categoryBETestResultMock } from '../../../shared/mocks/cat-be-test-result.mock';
```
Comment (mgoldson) [inline src/pages/view-test-result/cat-c/view-test-result.cat-c.page.ts]: Add TODO comment?

**Review by mgoldson — COMMENTED** [review]
Looking good. 
A few initial observations around adding TODOs worth looking at.
Will pull the branch and run later.

File: src/pages/view-test-result/cat-c/__tests__/view-test-result.cat-c.page.spec.ts (hunk 1 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/__tests__/view-test-result.cat-c.page.spec.ts:148]: Updated

File: src/pages/view-test-result/cat-c/components/business-details-card/__tests__/business-details-card.spec.ts (hunk 2 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/business-details-card/__tests__/business-details-card.spec.ts:47]: done

File: src/pages/view-test-result/cat-c/__tests__/view-test-result.cat-c.page.spec.ts (hunk 3 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/__tests__/view-test-result.cat-c.page.spec.ts:30]: it needs to be created have added a TODO - we don't have the schema yet

File: src/pages/view-test-result/cat-c/components/business-details-card/__tests__/business-details-card.spec.ts (hunk 4 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/business-details-card/__tests__/business-details-card.spec.ts:10]: done

File: src/pages/view-test-result/cat-c/components/debrief-card/__tests__/debrief-card.spec.ts (hunk 5 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/debrief-card/__tests__/debrief-card.spec.ts:24]: done

File: src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts (hunk 6 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts:27]: done

File: src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts (hunk 7 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts:97]: done

File: src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts (hunk 8 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts:13]: done

File: src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts (hunk 9 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/debrief-card/debrief-card.ts:6]: done

File: src/pages/view-test-result/cat-c/components/vehicle-details-card/__tests__/vehicle-details-card.spec.ts (hunk 10 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/vehicle-details-card/__tests__/vehicle-details-card.spec.ts:11]: done

File: src/pages/view-test-result/cat-c/components/vehicle-details-card/vehicle-details-card.ts (hunk 11 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/components/vehicle-details-card/vehicle-details-card.ts:3]: done

File: src/pages/view-test-result/cat-c/view-test-result.cat-c.page.module.ts (hunk 12 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/view-test-result.cat-c.page.module.ts]: fixed this it should be cat C

File: src/pages/view-test-result/cat-c/view-test-result.cat-c.page.ts (hunk 13 above)
Comment (TomBillingtonUK) [reply src/pages/view-test-result/cat-c/view-test-result.cat-c.page.ts]: done

**Pull request merged**